File: Atcoder_#310/javascript/c_reversible.js
function main(input) {
    const lines = input.trim().split('\n');
    const n = parseInt(lines[0], 10);
    const sticks = new Set();

    for (let i = 1; i <= n; i++) {
        const s = lines[i];
        const normalizedS = [s, s.split('').reverse().join('')].sort()[0];
        sticks.add(normalizedS);
    }

    // ユニークな種類の棒の数を出力
    console.log(sticks.size);
}

// 標準入力を読み込み、main関数を呼び出し
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin', 'utf8');
main(input);

File: Atcoder_#310/ruby/c_reversible.rb
def main
  n = gets.to_i
  sticks = []

  n.times do
    s = gets.chomp
    # 文字列 s とその反転のうち辞書順で小さい方を選ぶ
    normalized_s = [s, s.reverse].min
    sticks << normalized_s
  end

  # ユニークな文字列の個数を出力
  puts sticks.uniq.size
end

main
